File: 01_AI_Tech_Quest/backend_api/app/services/rag_service.py
from typing import Any

from app.data.mock_data import RAG_CHUNKS, RAG_DOCUMENT_TITLE
from app.schemas.rag import RagAskResponse, RagSourceResponse


FALLBACK_SOURCE = RagSourceResponse(
    source_id="Source 1",
    heading="未命中明確來源",
    chunk_id="fallback",
    snippet="目前 mock 文件沒有找到足夠相關的片段，正式 RAG 系統應該回答不知道或請使用者補充問題。",
    relevance=0.32,
)


def _normalize(text: str) -> str:
    return text.lower().strip()


def _score_chunk(question: str, chunk: Any) -> float:
    normalized_question = _normalize(question)
    score = 0.0

    for keyword in chunk.tags:
        if _normalize(keyword) in normalized_question:
            score += 2.5

    for token in _normalize(chunk.heading).split():
        if token and token in normalized_question:
            score += 1.0

    if chunk.heading in question:
        score += 1.5

    return score


def _build_answer(primary_heading: str, primary_body: str) -> str:
    return f"根據文件「{primary_heading}」，{primary_body}"


def ask_document(question: str, top_k: int = 2) -> RagAskResponse:
    trimmed_question = question.strip()

    ranked_chunks = sorted(
        ((_score_chunk(trimmed_question, chunk), chunk) for chunk in RAG_CHUNKS),
        key=lambda item: item[0],
        reverse=True,
    )
    matched_chunks = [(score, chunk) for score, chunk in ranked_chunks if score > 0]

    if not matched_chunks:
        return RagAskResponse(
            answer="目前文件中找不到足夠可靠的答案。正式的文件檢索增強生成（RAG）系統應該回答不知道，或請使用者提供更明確的問題。",
            confidence="low",
            sources=[FALLBACK_SOURCE],
            cited_snippets=[FALLBACK_SOURCE.snippet],
            retrieval_trace=f"mock 檢索未在「{RAG_DOCUMENT_TITLE}」找到高相關片段。",
        )

    selected_chunks = ranked_chunks[:top_k]
    best_chunk = selected_chunks[0][1]

    sources = [
        RagSourceResponse(
            source_id=f"Source {index + 1}",
            heading=chunk.heading,
            chunk_id=chunk.id,
            snippet=chunk.body,
            relevance=round(min(0.99, 0.42 + score * 0.1), 2),
        )
        for index, (score, chunk) in enumerate(selected_chunks)
    ]

    return RagAskResponse(
        answer=_build_answer(best_chunk.heading, best_chunk.body),
        confidence="high" if selected_chunks[0][0] >= 2.5 else "medium",
        sources=sources,
        cited_snippets=[source.snippet for source in sources],
        retrieval_trace=f"從「{RAG_DOCUMENT_TITLE}」取回 {len(sources)} 個相關文件片段，並用引用來源產生回答。",
    )


def ask_rag(question: str, top_k: int = 2) -> RagAskResponse:
    return ask_document(question=question, top_k=top_k)

File: 01_AI_Tech_Quest/frontend_web/src/features/rag-mission/services/ragService.js
import { ragDocument } from "../data/ragDocument.js";

const fallbackSource = {
  sourceId: "Source 1",
  heading: "未命中明確來源",
  chunkId: "fallback",
  snippet: "目前文件沒有找到足夠相關的片段，正式 RAG 系統應該回答不知道或請使用者補充問題。",
  relevance: 0.32,
};

function scoreChunk(question, chunk) {
  const normalized = question.toLowerCase();
  const tagScore = chunk.tags.reduce((score, tag) => {
    return normalized.includes(tag.toLowerCase()) ? score + 3 : score;
  }, 0);
  const headingTokens = chunk.heading.toLowerCase().split(/\s+/);
  const headingScore = headingTokens.reduce((score, token) => {
    return normalized.includes(token) ? score + 1 : score;
  }, 0);
  return tagScore + headingScore;
}

function buildAnswer(question, primaryChunk) {
  const answerByChunk = {
    "chunk-hours":
      "根據文件「營業時間」，晴宇咖啡平日營業時間為 10:00 到 20:00，週末與國定假日營業時間為 09:00 到 21:00。",
    "chunk-location":
      "根據文件「地址」，晴宇咖啡位於台北市晴宇路 100 號，距離最近捷運站步行約 5 分鐘。",
    "chunk-booking":
      "根據文件「預約方式」，顧客可以透過 LINE 預約座位；若當日仍有空位，也可以接受當日預約。",
    "chunk-power":
      "根據文件「插座資訊」，窗邊座位與共用工作桌大多提供插座，需要長時間使用筆電時建議預約插座座位。",
    "chunk-minimum":
      "根據文件「低消規則」，每位顧客低消為一杯飲品或一份甜點，停留超過兩小時建議追加點餐。",
  };

  return (
    answerByChunk[primaryChunk.id] ??
    `文件中有和「${question}」相關的內容，但 mock 檢索只找到部分相符片段。`
  );
}

export const ragService = {
  async ask(question) {
    const trimmedQuestion = question.trim();

    if (!trimmedQuestion) {
      throw new Error("請先輸入想詢問文件的問題。");
    }

    await new Promise((resolve) => {
      window.setTimeout(resolve, 650);
    });

    const rankedChunks = ragDocument.chunks
      .map((chunk) => ({
        chunk,
        score: scoreChunk(trimmedQuestion, chunk),
      }))
      .sort((a, b) => b.score - a.score);

    const matches = rankedChunks.filter((item) => item.score > 0);

    if (matches.length === 0) {
      return {
        answer:
          "目前文件中找不到足夠可靠的答案。正式的文件檢索增強生成（RAG）系統應該請使用者補充問題，或回答不知道。",
        confidence: "low",
        sources: [fallbackSource],
        citedSnippets: [fallbackSource.snippet],
        retrievalTrace: "mock 檢索沒有找到高相關文件片段。",
      };
    }

    const selectedSources = rankedChunks.slice(0, 2);
    const sources = selectedSources.map((item, index) => ({
      sourceId: `Source ${index + 1}`,
      heading: item.chunk.heading,
      chunkId: item.chunk.id,
      snippet: item.chunk.body,
      relevance: Math.min(0.97, 0.42 + item.score * 0.08),
    }));

    return {
      answer: buildAnswer(trimmedQuestion, matches[0].chunk),
      confidence: matches[0].score >= 3 ? "high" : "medium",
      sources,
      citedSnippets: sources.map((source) => source.snippet),
      retrievalTrace: `從「${ragDocument.title}」取回 ${sources.length} 個排序後的文件片段，主要回答依據最高相關來源。`,
    };
  },
};
